File: tinyagent/agent_tool_execution.py
"""Tool execution helpers for the agent loop."""

from __future__ import annotations

import asyncio
import json
from collections.abc import Awaitable, Callable

from pydantic import BaseModel, Field

from .agent_types import (
    AgentMessage,
    AgentTool,
    AgentToolResult,
    AssistantMessage,
    EventStream,
    JsonObject,
    MessageEndEvent,
    MessageStartEvent,
    TextContent,
    ToolCallContent,
    ToolExecutionEndEvent,
    ToolExecutionStartEvent,
    ToolExecutionUpdateEvent,
    ToolResultMessage,
)


class ToolExecutionResult(BaseModel):
    tool_results: list[ToolResultMessage] = Field(default_factory=list)
    steering_messages: list[AgentMessage] | None = None


def validate_tool_arguments(tool: AgentTool, tool_call: ToolCallContent) -> JsonObject:
    """Validate and normalize tool arguments.

    Some providers (e.g. MiniMax via alchemy) may return arguments as a JSON
    string rather than a parsed dict. Normalize to dict before execution.
    """

    del tool  # Arguments schema validation is deferred; normalize shape here.
    raw: object = tool_call.arguments

    if isinstance(raw, str):
        try:
            parsed = json.loads(raw) if raw else {}
            return parsed if isinstance(parsed, dict) else {}
        except (json.JSONDecodeError, ValueError):
            return {}

    return raw if isinstance(raw, dict) else {}


def _extract_tool_calls(assistant_message: AssistantMessage) -> list[ToolCallContent]:
    tool_calls: list[ToolCallContent] = []
    for content in assistant_message.content:
        if isinstance(content, ToolCallContent):
            tool_calls.append(content)
    return tool_calls


def _find_tool(tools: list[AgentTool] | None, name: str) -> AgentTool | None:
    if not tools:
        return None
    for tool in tools:
        if tool.name == name:
            return tool
    return None


def _is_parent_task_cancelling(parent_task: asyncio.Task[object] | None) -> bool:
    if parent_task is None:
        return False
    cancelling_attr = getattr(parent_task, "cancelling", None)
    if callable(cancelling_attr):
        cancelling_count = cancelling_attr()
        if isinstance(cancelling_count, int):
            return cancelling_count > 0
    return parent_task.cancelled()


async def _execute_single_tool(
    tool: AgentTool | None,
    tool_call: ToolCallContent,
    signal: asyncio.Event | None,
    stream: EventStream,
    parent_task: asyncio.Task[object] | None,
) -> tuple[AgentToolResult, bool]:
    """Execute a single tool and return (result, is_error)."""

    tool_call_name = tool_call.name or ""
    tool_call_id = tool_call.id or ""
    tool_call_args = tool_call.arguments

    if not tool:
        return (
            AgentToolResult(
                content=[TextContent(text=f"Tool {tool_call_name} not found")],
                details={},
            ),
            True,
        )
    if not tool.execute:
        error_text = f"Tool {tool_call_name} has no execute function"
        return (
            AgentToolResult(
                content=[TextContent(text=error_text)],
                details={},
            ),
            True,
        )

    try:
        validated_args = validate_tool_arguments(tool, tool_call)

        def on_update(partial_result: AgentToolResult) -> None:
            stream.push(
                ToolExecutionUpdateEvent(
                    tool_call_id=tool_call_id,
                    tool_name=tool_call_name,
                    args=tool_call_args,
                    partial_result=partial_result,
                )
            )

        result = await tool.execute(tool_call_id, validated_args, signal, on_update)
        return (result, False)
    except asyncio.CancelledError as exc:
        if _is_parent_task_cancelling(parent_task):
            raise
        message = str(exc) or "Tool execution cancelled"
        return (
            AgentToolResult(
                content=[TextContent(text=message)],
                details={},
            ),
            True,
        )
    except Exception as exc:  # noqa: BLE001
        return (
            AgentToolResult(
                content=[TextContent(text=str(exc))],
                details={},
            ),
            True,
        )


def _create_tool_result_message(
    tool_call: ToolCallContent,
    result: AgentToolResult,
    is_error: bool,
) -> ToolResultMessage:
    return ToolResultMessage(
        tool_call_id=tool_call.id or "",
        tool_name=tool_call.name or "",
        content=result.content,
        details=result.details,
        is_error=is_error,
        timestamp=int(asyncio.get_running_loop().time() * 1000),
    )


async def execute_tool_calls(
    tools: list[AgentTool] | None,
    assistant_message: AssistantMessage,
    signal: asyncio.Event | None,
    stream: EventStream,
    get_steering_messages: Callable[[], Awaitable[list[AgentMessage]]] | None = None,
) -> ToolExecutionResult:
    tool_calls = _extract_tool_calls(assistant_message)
    if not tool_calls:
        return ToolExecutionResult()

    # Emit start events for all tools upfront
    for tool_call in tool_calls:
        stream.push(
            ToolExecutionStartEvent(
                tool_call_id=tool_call.id or "",
                tool_name=tool_call.name or "",
                args=tool_call.arguments,
            )
        )

    # Resolve tools and execute all in parallel
    resolved = [_find_tool(tools, tc.name or "") for tc in tool_calls]
    parent_task = asyncio.current_task()
    raw_results: list[tuple[AgentToolResult, bool]] = await asyncio.gather(
        *(
            _execute_single_tool(tool, tc, signal, stream, parent_task)
            for tool, tc in zip(resolved, tool_calls, strict=True)
        )
    )

    # Emit end events and build result messages in original order
    results: list[ToolResultMessage] = []
    for tool_call, (result, is_error) in zip(tool_calls, raw_results, strict=True):
        stream.push(
            ToolExecutionEndEvent(
                tool_call_id=tool_call.id or "",
                tool_name=tool_call.name or "",
                result=result,
                is_error=is_error,
            )
        )
        tool_result_message = _create_tool_result_message(tool_call, result, is_error)
        results.append(tool_result_message)
        stream.push(MessageStartEvent(message=tool_result_message))
        stream.push(MessageEndEvent(message=tool_result_message))

    # Check for steering messages once after all tools complete
    steering_messages: list[AgentMessage] | None = None
    if get_steering_messages:
        steering = await get_steering_messages()
        if steering:
            steering_messages = steering

    return ToolExecutionResult(tool_results=results, steering_messages=steering_messages)


def skip_tool_call(tool_call: ToolCallContent, stream: EventStream) -> ToolResultMessage:
    """Skip a tool call due to user interruption."""

    tool_call_name = tool_call.name or ""
    tool_call_id = tool_call.id or ""
    tool_call_args = tool_call.arguments

    result = AgentToolResult(
        content=[TextContent(text="Skipped due to queued user message.")],
        details={},
    )
    stream.push(
        ToolExecutionStartEvent(
            tool_call_id=tool_call_id,
            tool_name=tool_call_name,
            args=tool_call_args,
        )
    )
    stream.push(
        ToolExecutionEndEvent(
            tool_call_id=tool_call_id,
            tool_name=tool_call_name,
            result=result,
            is_error=True,
        )
    )

    tool_result_message = _create_tool_result_message(tool_call, result, True)

    stream.push(MessageStartEvent(message=tool_result_message))
    stream.push(MessageEndEvent(message=tool_result_message))

    return tool_result_message

File: rust/src/agent_tool_execution.rs
use std::fmt;
use std::panic::AssertUnwindSafe;
use std::sync::Arc;
use std::time::{SystemTime, UNIX_EPOCH};

use futures::future::join_all;
use futures::FutureExt;
use tokio::sync::Notify;

use crate::agent_loop::AgentEventStream;
use crate::types::{
    AgentMessage, AgentMessageProvider, AgentTool, AgentToolResult, AgentToolUpdateCallback,
    AssistantContent, AssistantMessage, JsonObject, Message, MessageEndEvent, MessageStartEvent,
    TextContent, ToolCallContent, ToolExecutionEndEvent, ToolExecutionStartEvent,
    ToolExecutionUpdateEvent, ToolResultContent, ToolResultMessage,
};

pub type ToolExecutionContractResult<T> = Result<T, ToolExecutionError>;

#[derive(Debug, Clone, Default)]
pub struct ToolExecutionResult {
    pub tool_results: Vec<ToolResultMessage>,
    pub steering_messages: Option<Vec<AgentMessage>>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ToolExecutionError {
    MissingToolCallId,
    MissingToolCallName,
}

impl fmt::Display for ToolExecutionError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::MissingToolCallId => f.write_str("tool call is missing required id"),
            Self::MissingToolCallName => f.write_str("tool call is missing required name"),
        }
    }
}

impl std::error::Error for ToolExecutionError {}

#[derive(Debug, Clone)]
struct ExecutableToolCall {
    id: String,
    name: String,
    arguments: JsonObject,
}

pub fn validate_tool_arguments(tool: &AgentTool, tool_call: &ToolCallContent) -> JsonObject {
    let _ = tool;
    tool_call.arguments.clone()
}

pub async fn execute_tool_calls(
    tools: Option<&[AgentTool]>,
    assistant_message: &AssistantMessage,
    signal: Option<Arc<Notify>>,
    stream: &AgentEventStream,
    get_steering_messages: Option<AgentMessageProvider>,
) -> ToolExecutionContractResult<ToolExecutionResult> {
    let tool_calls = extract_tool_calls(assistant_message)?;
    if tool_calls.is_empty() {
        return Ok(ToolExecutionResult::default());
    }

    for tool_call in &tool_calls {
        stream.push(crate::types::AgentEvent::ToolExecutionStart(
            ToolExecutionStartEvent {
                tool_call_id: tool_call.id.clone(),
                tool_name: tool_call.name.clone(),
                args: Some(tool_call.arguments.clone()),
                ..Default::default()
            },
        ));
    }

    let resolved_tools = tool_calls
        .iter()
        .map(|tool_call| find_tool(tools, &tool_call.name).cloned())
        .collect::<Vec<_>>();

    let raw_results = join_all(
        resolved_tools
            .into_iter()
            .zip(tool_calls.iter().cloned())
            .map(|(tool, tool_call)| {
                execute_single_tool(tool, tool_call, signal.clone(), Arc::clone(stream))
            }),
    )
    .await;

    let mut tool_results = Vec::with_capacity(raw_results.len());

    for (tool_call, (result, is_error)) in tool_calls.iter().zip(raw_results.into_iter()) {
        stream.push(crate::types::AgentEvent::ToolExecutionEnd(
            ToolExecutionEndEvent {
                tool_call_id: tool_call.id.clone(),
                tool_name: tool_call.name.clone(),
                result: Some(result.clone()),
                is_error,
                ..Default::default()
            },
        ));

        let tool_result_message = create_tool_result_message(tool_call, result, is_error);
        let tool_result_agent_message =
            AgentMessage::Message(Message::ToolResult(tool_result_message.clone()));

        tool_results.push(tool_result_message);
        stream.push(crate::types::AgentEvent::MessageStart(MessageStartEvent {
            message: Some(tool_result_agent_message.clone()),
            ..Default::default()
        }));
        stream.push(crate::types::AgentEvent::MessageEnd(MessageEndEvent {
            message: Some(tool_result_agent_message),
            ..Default::default()
        }));
    }

    let steering_messages = match get_steering_messages {
        Some(provider) => {
            let messages = provider().await;
            if messages.is_empty() {
                None
            } else {
                Some(messages)
            }
        }
        None => None,
    };

    Ok(ToolExecutionResult {
        tool_results,
        steering_messages,
    })
}

pub fn skip_tool_call(
    tool_call: &ToolCallContent,
    stream: &AgentEventStream,
) -> ToolExecutionContractResult<ToolResultMessage> {
    let executable_tool_call = executable_tool_call(tool_call)?;
    let result = AgentToolResult {
        content: vec![ToolResultContent::Text(TextContent {
            text: Some("Skipped due to queued user message.".to_string()),
            ..Default::default()
        })],
        details: JsonObject::new(),
    };

    stream.push(crate::types::AgentEvent::ToolExecutionStart(
        ToolExecutionStartEvent {
            tool_call_id: executable_tool_call.id.clone(),
            tool_name: executable_tool_call.name.clone(),
            args: Some(executable_tool_call.arguments.clone()),
            ..Default::default()
        },
    ));
    stream.push(crate::types::AgentEvent::ToolExecutionEnd(
        ToolExecutionEndEvent {
            tool_call_id: executable_tool_call.id.clone(),
            tool_name: executable_tool_call.name.clone(),
            result: Some(result.clone()),
            is_error: true,
            ..Default::default()
        },
    ));

    let tool_result_message = create_tool_result_message(&executable_tool_call, result, true);
    let tool_result_agent_message =
        AgentMessage::Message(Message::ToolResult(tool_result_message.clone()));

    stream.push(crate::types::AgentEvent::MessageStart(MessageStartEvent {
        message: Some(tool_result_agent_message.clone()),
        ..Default::default()
    }));
    stream.push(crate::types::AgentEvent::MessageEnd(MessageEndEvent {
        message: Some(tool_result_agent_message),
        ..Default::default()
    }));

    Ok(tool_result_message)
}

async fn execute_single_tool(
    tool: Option<AgentTool>,
    tool_call: ExecutableToolCall,
    signal: Option<Arc<Notify>>,
    stream: AgentEventStream,
) -> (AgentToolResult, bool) {
    match tool {
        None => (
            AgentToolResult {
                content: vec![ToolResultContent::Text(TextContent {
                    text: Some(format!("Tool {} not found", tool_call.name)),
                    ..Default::default()
                })],
                details: JsonObject::new(),
            },
            true,
        ),
        Some(tool) => match tool.execute {
            None => (
                AgentToolResult {
                    content: vec![ToolResultContent::Text(TextContent {
                        text: Some(format!("Tool {} has no execute function", tool_call.name)),
                        ..Default::default()
                    })],
                    details: JsonObject::new(),
                },
                true,
            ),
            Some(ref execute) => {
                let tool_call_id = tool_call.id.clone();
                let tool_call_name = tool_call.name.clone();
                let tool_call_args = tool_call.arguments.clone();
                let update_stream = Arc::clone(&stream);
                let tool_call_content = ToolCallContent {
                    id: Some(tool_call.id.clone()),
                    name: Some(tool_call.name.clone()),
                    arguments: tool_call.arguments.clone(),
                    ..Default::default()
                };
                let validated_args = validate_tool_arguments(&tool, &tool_call_content);
                let on_update: AgentToolUpdateCallback = Arc::new(move |partial_result| {
                    update_stream.push(crate::types::AgentEvent::ToolExecutionUpdate(
                        ToolExecutionUpdateEvent {
                            tool_call_id: tool_call_id.clone(),
                            tool_name: tool_call_name.clone(),
                            args: Some(tool_call_args.clone()),
                            partial_result: Some(partial_result),
                            ..Default::default()
                        },
                    ));
                });

                let future = execute(tool_call.id.clone(), validated_args, signal, on_update);

                match AssertUnwindSafe(future).catch_unwind().await {
                    Ok(result) => (result, false),
                    Err(_) => (
                        AgentToolResult {
                            content: vec![ToolResultContent::Text(TextContent {
                                text: Some(format!("Tool {} panicked during execution", tool_call.name)),
                                ..Default::default()
                            })],
                            details: JsonObject::new(),
                        },
                        true,
                    ),
                }
            }
        },
    }
}

fn extract_tool_calls(
    assistant_message: &AssistantMessage,
) -> ToolExecutionContractResult<Vec<ExecutableToolCall>> {
    assistant_message
        .content
        .iter()
        .filter_map(|content| match content {
            Some(AssistantContent::ToolCall(tool_call)) => Some(executable_tool_call(tool_call)),
            _ => None,
        })
        .collect()
}

fn executable_tool_call(
    tool_call: &ToolCallContent,
) -> ToolExecutionContractResult<ExecutableToolCall> {
    let id = tool_call
        .id
        .clone()
        .ok_or(ToolExecutionError::MissingToolCallId)?;
    let name = tool_call
        .name
        .clone()
        .ok_or(ToolExecutionError::MissingToolCallName)?;

    Ok(ExecutableToolCall {
        id,
        name,
        arguments: tool_call.arguments.clone(),
    })
}

fn find_tool<'a>(tools: Option<&'a [AgentTool]>, name: &str) -> Option<&'a AgentTool> {
    tools.and_then(|tools| tools.iter().find(|tool| tool.name == name))
}

fn create_tool_result_message(
    tool_call: &ExecutableToolCall,
    result: AgentToolResult,
    is_error: bool,
) -> ToolResultMessage {
    ToolResultMessage {
        tool_call_id: Some(tool_call.id.clone()),
        tool_name: Some(tool_call.name.clone()),
        content: result.content,
        details: result.details,
        is_error,
        timestamp: Some(current_timestamp_ms()),
        ..Default::default()
    }
}

fn current_timestamp_ms() -> i64 {
    match SystemTime::now().duration_since(UNIX_EPOCH) {
        Ok(duration) => duration.as_millis() as i64,
        Err(error) => -(error.duration().as_millis() as i64),
    }
}
